### ISSUE TITLE
Add mypy-lang to static code analysis?

### ISSUE THREAD

**Issue by carlthome** [spyder-ide/spyder#3045]
It would be wonderful to have static typing warnings via mypy-lang pop up directly in the IDE like the current static code analysis does. Is this a planned feature?

**Comment by ccordoba12** [comment]
It's not in our roadmap, sorry. Maybe this could be implemented as a plugin, after Spyder 3.0 is released (in a couple of months :-)

**Comment by Nodd** [comment]
It's a big thing for the next python version so it should be a priority after 3.0.

**Comment by RnoldR** [comment]
Any chance to have MyPy implemented in version 4.0? Could not find any reference to it, would be a great addition to the linter section. Running MyPy from the command line is really tedious.

**Comment by goanpeca** [comment]
@RnoldR spyder for now uses the Language Server Protocol for the completion archetecture and can use plugins written for it like:

https://github.com/tomv564/pyls-mypy

So in theory there is support out of the box by installing that package

@ccordoba12 should we consider including this as a dependency?

**Comment by ccordoba12** [comment]
> should we consider including this as a dependency?

We need to see if that plugin flags things without typing as warnings/errors because that would be very confusing to newbies. If not, then I agree to add it as another dependency.

**Comment by ccordoba12** [comment]
@steff456, please install pyls-mypy and check how things look with it. Thanks!

**Comment by RnoldR** [comment]
Great for looking into this!  Agree that untyped variables should not be flagged, which is also required by PEP 484. Some quick tests indicate that mypy does not flag untyped variables.

**Comment by RnoldR** [comment]
Let me know if I can test.

**Comment by ccordoba12** [comment]
If you're using pip you can test this right now by running

    pip install pyls-mypy

**Comment by RnoldR** [comment]
It works! I'll start experimenting right away.

**Comment by RnoldR** [comment]
Oh, an oversight: I use Anaconda but it still works.

**Comment by RnoldR** [comment]
But it broke my installation, most packages are gone. Fortunately that only happened in the test environment. Other environments seem to be untouched.

**Comment by ccordoba12** [comment]
> but it still works

Great!! @steff456 will do more tests, and if everything goes well, we could provide out of the box support for MyPy in our next beta.

**Comment by RnoldR** [comment]
That would make me very happy! MyPy is a great addition, it really makes my code better.

**Comment by steff456** [comment]
@ccordoba12 I already try it in an environment created with pip, not conda and it is working fine and the environment is not broken.

**Comment by RnoldR** [comment]
I have tested more thoroughly and it seems to work now, I had some issues with my spyder 4 install which I resolved and it all seems to work fine. I'll continue rewriting my code and will report if something is found. Thanks for all your efforts!

**Comment by ccordoba12** [comment]
@RnoldR, yes, please let us know if you encounter any issues so we can be more confident when adding that package as one more of our dependencies.

**Comment by dfrison01** [comment]
Just tested on Python 3.7.4 ; Spyder 4.1.3 and pyls-mypy 0.1.8. Works like a charm. And I recently discovered that even Cython benefits from type annotations too. Thank you very much, I discovered all that just because of this ticket. Great addition!

**Comment by ccordoba12** [comment]
Thanks! We'll provide a simpler experience to enable and work with MyPy in our 4.2.0 version, to be released in a couple of months.

**Comment by lundybernard** [comment]
@Greedquest pylsp-mypy is currently working on the latest version of conda

conda version : 4.11.0
python version : 3.9.7.final.0

# Name                    Version                   Build  Channel
mypy                      0.931                    pypi_0    pypi
mypy_extensions           0.4.3            py39h06a4308_1  
pylsp-mypy                0.5.7              pyhd8ed1ab_0    conda-forge
pyls-spyder               0.4.0              pyhd3eb1b0_0  
spyder                    5.1.5            py39h06a4308_1  
spyder-kernels            2.1.3            py39h06a4308_0  

Native support would be excellent, and I hope to see some movement on #13360 soon

**Comment by Greedquest** [comment]
@lundybernard yes, actually I got it working too. At the time I wrote that comment, pylsp-mypy (the fork) had upgraded relative to pyls-mypy to use a newer language server backend - but Spyder was still using the old language server. Recent versions of Spyder have now caught up and use the new language server so are compatible with pylsp-mypy (and incompatible with pyls-mypy)

Or in simpler terms; for a short time the mypy plug-in and Spyder were out if step, but now they're back working well together again.
